compiler/noirc_evaluator/src/ssa/opt/inline_const_brillig_calls.rs

```diff
@@ -0,0 +1,220 @@
+                continue;
+            }
+
+            // We also don't want to remove entry points
+            if self.entry_point_to_generated_index.contains_key(func_id) {
+                continue;
+            }
+
+            self.functions.remove(func_id);
+        }
+
+        self
+    }
+}
+
+impl Function {
+    pub(crate) fn inline_const_brillig_calls(
+        &mut self,
+        brillig_functions: &BTreeMap<FunctionId, Function>,
+        brillig_functions_we_could_not_inline: &mut HashSet<FunctionId>,
+        inliner_aggressiveness: i64,
+        error_selector_to_type: &BTreeMap<ErrorSelector, Type>,
+    ) {
+        for block_id in self.reachable_blocks() {
+            for instruction_id in self.dfg[block_id].take_instructions() {
+                if !self.optimize_const_brillig_call(
+                    instruction_id,
+                    brillig_functions,
+                    brillig_functions_we_could_not_inline,
+                    inliner_aggressiveness,
+                    error_selector_to_type,
+                ) {
+                    self.dfg[block_id].instructions_mut().push(instruction_id);
+                }
+            }
+        }
+    }
+
+    /// Tries to optimize an instruction if it's a call that points to a brillig function,
+    /// and all its arguments are constant. If the optimization is successful, returns true.
+    /// Otherwise returns false. The given instruction is not removed from the function.
+    fn optimize_const_brillig_call(
+        &mut self,
+        instruction_id: InstructionId,
+        brillig_functions: &BTreeMap<FunctionId, Function>,
+        brillig_functions_we_could_not_inline: &mut HashSet<FunctionId>,
+        inliner_aggressiveness: i64,
+        error_selector_to_type: &BTreeMap<ErrorSelector, Type>,
+    ) -> bool {
+        let instruction = &self.dfg[instruction_id];
+        let Instruction::Call { func: func_id, arguments } = instruction else {
+            return false;
+        };
+
+        let func_value = &self.dfg[*func_id];
+        let Value::Function(func_id) = func_value else {
+            return false;
+        };
+
+        let Some(function) = brillig_functions.get(func_id) else {
+            return false;
+        };
+
+        if !arguments.iter().all(|argument| self.dfg.is_constant(*argument)) {
+            brillig_functions_we_could_not_inline.insert(*func_id);
+            return false;
+        }
+
+        // The function we have is already a copy of the original function, but we need to clone
+        // it again because there might be multiple calls to the same brillig function.
+        let mut function = Function::clone_with_id(*func_id, function);
+
+        // Find the entry block and remove its parameters
+        let entry_block_id = function.entry_block();
+        let entry_block = &mut function.dfg[entry_block_id];
+        let entry_block_parameters = entry_block.take_parameters();
```

Comment:
I suppose this is another reason it must be cloned, otherwise we might be mutating the original block. 

I actually don't understand why this works: in `Function::inline_const_brillig_calls` above we started with `self.dfg[block_id].take_instructions()` of all `self.reachable_blocks()`; can't the `entry_block` already have its parameters taken?

Reply:
Ah it's `take_parameters` not `take_instructions` 🥹